core/src/main/java/org/elasticsearch/action/update/UpdateHelper.java

```diff
@@ -298,8 +304,24 @@ public GetResult extractGetResult(final UpdateRequest request, String concreteIn
             }
         }
 
+        BytesReference sourceFilteredAsBytes = sourceAsBytes;
+        if (request.fetchSource() != null) {
```

Comment:
Is it worth optimizing the case where `fetchSource` is true and `includes` and `excludes` are empty and the xcontents line up? Then you can just return the bytes you have like we do for search lookups.